### PR TITLE
feat: migrate to Yew 0.20, using yew-nested-router in the process

### LINKED ISSUE #5: Add support for the Next (after 0.19) version of Yew
I really like this lib, makes things so much easier!

I'm not able to use this component with the latest version of Yew. I know it isn't released yet, but I was hoping to upgrade my app to that in preparation for its release.

The error I'm getting is:

```
34 | pub struct OAuth2<C: Client> {
   | ---------------------------- doesn't satisfy `_: yew::BaseComponent`
   |
   = note: the following trait bounds were not satisfied:
           `yew_oauth2::components::context::OAuth2<OAuth2Client>: yew::BaseComponent`
```

I understand if the Yew code base is too fluid at the moment, so happy if it sits here until it's closer to release. I might come back and try and do it myself.

### PR DESCRIPTION
**Pull request by ctron**
This rewrites some of the internals, as Yew dropped the support for Agents.

closes #5

### CODE REVIEW

**Comment by stephenandary** [comment]
So the main changes for this upgrade are using the new nested router crate which you created, leveraging the original style of router? That currently does not have history, but that should be okay if using something like yewdux right?

File: Cargo.toml [hunk 1]
```diff
@@ -21,20 +21,20 @@ num-traits = "0.2"
 oauth2 = "4"
 reqwest = "0.11"
 serde = { version = "1", features = ["derive"] }
+tokio = { version = "1", features = ["sync"] }
 wasm-bindgen = "0.2"
 wasm-bindgen-futures = "0.4"
-yew = "0.19"
-yew-agent = "0.1"
+yew = "0.20"
```
Comment (stephenandary) [inline Cargo.toml:27]: I think you need to put "0.20.0", This is what I have in my yew app to get renderer to work. `yew = { version = "0.20.0", features = ["csr"] }`

Comment (ctron) [reply Cargo.toml:27]: True, you need that for an app, but not for a library.

**Comment by ctron** [comment]
> > This rewrites some of the internals, as Yew dropped the support for Agents.
> > closes #5
> 
> What are agents used for? Why did they drop support for them?

There was an agent, managing the internal state. The agent was basically and actor pattern (similar to actix) that made it easier to maintain a mutable state.

The change now is to basically re-implement the necessary bits in this crate.

`yew-agent` supported two modes of operation: "context" and "web workers". Context was running in the single browser thread, leveraging the "context" idea of Yew (similar to the components). Web workers are far more complex.

The argument is, that the "context" style agents are now no longer necessary as Yew bring the new "context" system (`use_context` and `<ContextProvider<T>>`. And there is a general push for "function components".

**Comment by ctron** [comment]
> So the main changes for this upgrade are using the new nested router crate which you created, leveraging the original style of router? That currently does not have history, but that should be okay if using something like yewdux right?

The two main changes are: replace the `yew-agent` based Agent with a custom wrapper, which offers the necessary decoupling to let the OAuth2 state run in an agent-style fashion, but fully tie it in to the new `use_context` style.

Second, to replace the (optional) router integration with the new `yew-nested-router`. Still this is optional.

The current implementation leverages the browser history, so just pressing the back button on the browser should take you back to the previous route (not the previous page).

File: Cargo.toml (hunk 1 above)
Comment (stephenandary) [reply Cargo.toml:28]: Ah, makes sense!

**Comment by stephenandary** [comment]
This is so cool, thank you for sharing your knowledge! Do you have a Slack channel for your projects? It would be awesome to chat outside of PR

**Comment by ctron** [comment]
I am mostly on Matrix these days (@ctron:dentrassi.de). I can create a room there is you think this makes sense.

**Comment by stephenandary** [comment]
> I am mostly on Matrix these days (@ctron:dentrassi.de). I can create a room there is you think this makes sense.

Very cool, that would be great! Thank you

**Comment by ctron** [comment]
> > I am mostly on Matrix these days (@ctron:dentrassi.de). I can create a room there is you think this makes sense.
> 
> Very cool, that would be great! Thank you

Any suggestion for a name that fits? `yew-oauth2` sounds plain boring :D

**Comment by stephenandary** [comment]
> > > I am mostly on Matrix these days (@ctron:dentrassi.de). I can create a room there is you think this makes sense.
> > 
> > Very cool, that would be great! Thank you
> 
> Any suggestion for a name that fits? `yew-oauth2` sounds plain boring :D

What about yew-authorized (like "you are authorized" slang, and because it does openid connect too? Haha) Still boring?

**Comment by ctron** [comment]
I created `#yew-tools:matrix.org`, which might also cover the patternfly-yew project and the router stuff.

**Pull request merged**